discovery-utils/src/registration_client.rs

```diff
@@ -48,3 +56,70 @@ pub async fn register_discovery_handler_again(
         }
     }
 }
+
+
+//After discovery handler finds a list of devices. Discovery handler can use query_devices to fetch additional device information and put that into Device properties
+#[derive(Clone)]
+pub struct DeviceQueryInput {
+    pub id: String,
+    pub properties: HashMap<String,String>,
+    pub query_device_payload: Option<String>,
+    pub mounts: Vec<Mount>,
+}
+
+async fn query_device_info(
+     mut query_input: DeviceQueryInput, query_http: String
+) -> Device {
+    let result = get_client().await;
+    if let Ok(mut client) = result{
+        if let Some(query_payload)=query_input.query_device_payload {
+            let request = Request::new(QueryDeviceInfoRequest{
+                query_device_payload:query_payload,
+                query_device_http:query_http
+            });
+        
+            let result = client.query_device_info(request).await;
+            if let Ok(query_device_response) = result {
+                let device_ext_info=query_device_response.into_inner().query_device_result;
+                if device_ext_info.len()>0 {
+                    query_input.properties.insert(crate::DEVICE_EXT_INFO_LABEL.to_string(),device_ext_info);
+                }
+            }
+        }
+    }
+
+    Device {
+        id: query_input.id,
+        properties:query_input.properties,
+        mounts: query_input.mounts,
+        device_specs: Vec::default(),
+    }
+    
+}
+
+pub async fn query_devices(devices_query:Vec<DeviceQueryInput>,query_device_http:Option<String>) -> Vec<Device>{
+    // in case there is query http for asking for additional device information, raise multiple rpc calls to query akri agent
+    if let Some(query_http)=query_device_http{
+        let mut fetch_devices_tasks = Vec::new();
+        for device_query in devices_query.clone() {
+            fetch_devices_tasks.push(tokio::spawn(query_device_info(device_query,query_http.clone())));    
+        }
+        let result= futures::future::try_join_all(fetch_devices_tasks).await;
+        if let Ok(query_devices_result) = result{
+            return query_devices_result;
+        }
+    }
+    
+    //if there is no need to query additional device information or the previous device query is not completely successful, simply assemeble Devices and return
```

Comment:
`is not complteted successful` ... why no error proprogation here?  should we at least log the failure?

Reply:
Http query should be logged. I will add it into query_device_info(registration_client.rs) .
All discovered device will be used to query info. There is chance some queries may fail. In current implementation, query_device_info will be invoked by each device. It will simply give "Device" result if a device query fails and wont' propagate to fail the future try_join_all call. Let's say 10 devices are queried and one fail, we will still get 9 correct assembled "Device" and the 10th "Device" with empty external information.